File: ecommerce-frontend/src/app/services/user.service.ts
import { Injectable } from '@angular/core';
import { HttpClient } from '@angular/common/http';
import { Observable, tap } from 'rxjs';

/**
 * 用户服务
 * 封装用户注册、登录、获取用户信息等 API 调用
 * 管理 JWT token 的存储和读取
 */
@Injectable({
  providedIn: 'root'
})
export class UserService {

  /** token 在 localStorage 中的存储键名 */
  private readonly TOKEN_KEY = 'token';

  constructor(private http: HttpClient) {}

  /**
   * 用户注册
   * @param username 用户名
   * @param password 密码
   * @param email 邮箱
   * @returns 包含用户ID的响应
   */
  register(username: string, password: string, email: string): Observable<any> {
    return this.http.post('/api/user/register', { username, password, email });
  }

  /**
   * 用户登录
   * 登录成功后自动将 token 保存到 localStorage
   * @param username 用户名
   * @param password 密码
   * @returns 包含 JWT token 的响应
   */
  login(username: string, password: string): Observable<any> {
    return this.http.post('/api/user/login', { username, password }).pipe(
      tap((res: any) => {
        // 登录成功后将 token 保存到 localStorage
        if (res && res.data) {
          localStorage.setItem(this.TOKEN_KEY, res.data);
        }
      })
    );
  }

  /**
   * 获取当前登录用户信息
   * @returns 包含用户信息的响应
   */
  getUserInfo(): Observable<any> {
    return this.http.get('/api/user/info');
  }

  /**
   * 从 localStorage 获取 token
   * @returns token 字符串，不存在则返回 null
   */
  getToken(): string | null {
    return localStorage.getItem(this.TOKEN_KEY);
  }

  /**
   * 检查用户是否已登录
   * @returns 是否存在有效 token
   */
  isLoggedIn(): boolean {
    return !!this.getToken();
  }

  /**
   * 退出登录
   * 清除 localStorage 中的 token
   */
  logout(): void {
    localStorage.removeItem(this.TOKEN_KEY);
  }
}

File: ecommerce-service/src/main/java/com/ecommerce/service/UserService.java
package com.ecommerce.service;

import com.ecommerce.model.entity.User;

/**
 * 用户服务接口
 * 提供用户注册、登录、查询等功能
 */
public interface UserService {

    /**
     * 用户注册
     * @param username 用户名
     * @param password 密码（明文）
     * @param email 邮箱
     * @return 新创建的用户ID
     */
    Long register(String username, String password, String email);

    /**
     * 用户登录
     * @param username 用户名
     * @param password 密码（明文）
     * @return JWT token
     */
    String login(String username, String password);

    /**
     * 根据ID查询用户
     * @param id 用户ID
     * @return 用户实体
     */
    User getUserById(Long id);

    /**
     * 根据用户名查询用户
     * @param username 用户名
     * @return 用户实体，不存在返回 null
     */
    User getUserByUsername(String username);
}
